### ISSUE TITLE
Forms demo 

### ISSUE THREAD

**Issue by LHinson** [patternfly/patternfly-next#200]
This is related to #289 and should be done in parallel.
 
A forms demo is comprised of the form component (basic form elements see #26) plus any additional features to enhance form examples.  

Requirements: https://docs.google.com/document/d/1CRkRwx-GWjaVMA6HvoCbFe0CIqGCxBrQMfn8eshDvQ0/edit?usp=sharing

[Latest forms design PDF](https://drive.google.com/drive/folders/19dX04YdM2rnI00Eh-fOR4CXFt5rIq3i8)

Component lifecycle: 
- [x] Interaction design requirements 
- [x] Visuals 
- [ ] First checkpoint complete and approved 
- [ ] CSS 
- [ ] React
- [ ] Second checkpoint complete and approved 

Contacts: 
- IxD: @mcarrano 
- Visual 
- CSS

**Comment by mcarrano** [comment]
@LHinson I added a link to my initial requirements/discovery document in the description. This mostly summarizes the features of forms as available today in PF 3. Since we've recently gone through a [detailed analysis of Forms](https://docs.google.com/document/d/1PcxszZ1S53Oacf7XZddnvba42RjMGr5R_QtQHZUqq98/edit?usp=sharing), I don't think a lot of discovery work is needed here. Let me know if you agree. Should be ready to pass on to visual design.

**Comment by mcarrano** [comment]
@sjcox-rh as part of this work, please consider this issue that was filed against PatternFly 3 https://github.com/patternfly/patternfly-design/issues/671. It concerns vertical spacing of fields and whether we should leave space for displaying possible error messages without growing the form.

**Comment by kybaker** [comment]
Added the latest forms PDF above.

**Comment by catrobson** [comment]
Two things that stuck out to me:

* The right aligned button field - this didn't feel like a button at first glance, and I was also concerned if it will get lost with the active state on the field.  
![image](https://user-images.githubusercontent.com/3739227/42058095-4bb7bc3e-7aed-11e8-9cda-612ad9ccac27.png)

* The multi-select field.  There are a lot of different requirements around this capability from the products I think, and since we never had anything beyond the bootstrap combobox multi-select specified in PF3 I think this might warrant a bit more IXD time to work through all the states and behaviors expected.

Otherwise the design looks great!

**Comment by thomasmaas** [comment]
Looking good. Some issues I see:

- Not sure how needed/practical/nice the change in border width will be for states: it means we are occupying more/less vertical space but at the same time we don't want other element to start jumping so we will have to diminish the vertical spacing between the element in the different state and it's siblings. 
- I'm wondering how the vertical rhythm is designed; I would expect the submit button to have the same height as an input element. 
- Would be good to work out the submit button placement a bit further in different use cases. In PF3 the submit / cancel button locations depend on the use case. I would prefer a solution where the submit button always sits at the same place. Imho that should be on the right with the cancel link on the left.

**Comment by mcarrano** [comment]
@thomasmaas The reasoning for placing the submit button to the left on a full page form is that as the user scans down the form, the primary action is aligned with the left edges of the other controls. It would not seem right to have the Cancel button aligned this way IMO. @mceledonia what are your thoughts regarding this and other issues raised by @thomasmaas ?

**Comment by thomasmaas** [comment]
@mcarrano I understand the reasoning behind the left alignment but a full page form will likely have a limited visual width so a right aligned submit button wouldn't be that far from the vertical scan line. Plus a lot can be done by visually making the submit button stronger than the cancel link (which shouldn't look like a button in my book). Take this Github comment form, it's "full width" and the confirm/submit/comment button sits on the right. Right-side expresses: next. The next button of a wizard also makes more sense at the right. 

I guess my main motivation is consistency: to have 1 place where a user can expect that submit/next/confirm button to be and the logical place for me would be the right side, where the enter key sits on a keyboard.

**Comment by mcarrano** [comment]
@thomasmaas I think what you are saying makes sense. With PatternFly 3, the problem was that forms did not have a fixed width, so when we aligned buttons to the right on forms and wizards we ended up with weird layouts where the submit buttons would be far from the content for full page forms and wizards.

I do agree if we can constrain the width of a form for PatternFly 4, it will leave us in a better place. I think this is an ongoing discussion we need with the CSS team regarding gridding and layout.
